viewing_party/party.py

```diff
@@ -1,23 +1,197 @@
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+
+def get_unique_watched(user_data):
+    user_watched_list = []
+    friend_watched_list = []
+
+    for movie in user_data["watched"]:
+        user_watched_list.append(movie["title"])
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friend_watched_list.append(movie["title"])
+    
+    user_watched_set = set(user_watched_list)
+    friend_watched_set = set(friend_watched_list)
+    user_unique_watched_set = user_watched_set.difference(friend_watched_set)
+    user_unique_watched_list = []
+
+    for unique_movie in user_unique_watched_set:
+        for movie in user_data["watched"]:
+            if movie["title"] == unique_movie:
+                user_unique_watched_list.append(movie)
+
+    return user_unique_watched_list
+
+def get_friends_unique_watched(user_data):
+    user_watched_list = []
+    friend_watched_list = []
+
+    for movie in user_data["watched"]:
+        user_watched_list.append(movie["title"])
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friend_watched_list.append(movie["title"])
+    
+    user_watched_set = set(user_watched_list)
+    friend_watched_set = set(friend_watched_list)
+    friend_unique_watched_set = friend_watched_set.difference(user_watched_set)
+    friend_unique_watched_list = []
+
+    for unique_movie in friend_unique_watched_set:
+        for friend in user_data["friends"]:
+            for movie in friend["watched"]:
+                if movie not in friend_unique_watched_list:
+                    if movie["title"] == unique_movie:
+                        friend_unique_watched_list.append(movie)
+
+    return friend_unique_watched_list
+
+
+
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+def get_available_recs(user_data):
+
+    friends_unique_watched_list = get_friends_unique_watched(user_data)
+
+    available_recommended_movies_list = []
+
+    for movie in friends_unique_watched_list:
+        for subscription in user_data["subscriptions"]:
+            if subscription == movie["host"]:
+                available_recommended_movies_list.append(movie)
+
+    return available_recommended_movies_list
+
+
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+def get_new_rec_by_genre(user_data):
+
+    most_watched_genre = get_most_watched_genre(user_data)
+    friend_unique_watched = get_friends_unique_watched(user_data)
+    recommended_movies = []
+
+    for movie in friend_unique_watched:
+        if movie not in recommended_movies and movie["genre"] == most_watched_genre:
+            recommended_movies.append(movie)
+    return recommended_movies
+
+def get_rec_from_favorites(user_data):

```

Comment:
👍 

There are a few `in` checks here which we could think about addressing either by reorganizing the order of the looping, or by making helper sets of titles as described above.